Enabling USB2SNES Support

MattEqualsCoder opened the pull request:
This change is to add USB2SNES / QUSB2SNES support for auto tracking.

Unfortunately, QUSB2SNES works entirely differently from the Lua scripts. For the Lua scripts Tracker actually acts as the host of a regular socket server, but the QUSB2SNES app actually runs a websocket server that Tracker needs to act as a client to.

Because of this, I broke the connectors out into separate classes based off of the same interface and switch between the two when the player changes auto tracking connections and inform the auto tracker of new messages via events. I also pulled out the actual Auto Tracking into its own class instead of being contained in the module class so that when voice commands start getting added the module code itself will be cleaner.

Vivelin commented on src/Randomizer.App/ViewModels/GeneralOptions.cs:
```diff
@@ -67,7 +82,7 @@ public static IEnumerable<string> QuickLaunchOptions
         public bool TrackerSpoilersEnabled { get; set; }
 
-        public bool AutoTrackerAutoStart { get; set; }
+        public int AutoTrackerConnectorType { get; set; } = (int)EmulatorConnectorType.None;
```
Why is this an int?

MattEqualsCoder replied:
Because unfortunately, last I knew, C# combo boxes don't work super well being based off of enums with descriptions, so one of the ways around that was to select the descriptions as strings and use them as the value and bind the selected index back, which is then casted by to the same enum used to build the combo box.

I could look into creating an extension or type converter for the enum if that would be preferred.

Vivelin commented on src/Randomizer.App/ViewModels/RandomizerOptions.cs:
```diff
@@ -81,9 +82,9 @@ public string AutoTrackerScriptsOutputPath
         }
 
-        public bool AutotrackerAutoStart
+        public EmulatorConnectorType AutotrackerAutoStart
```
This should be either renamed or explained.

Vivelin commented on src/Randomizer.SMZ3.Tracking/AutoTracking/AutoTracker.cs:
```diff
@@ -0,0 +1,695 @@
+                }
+                catch (Exception e)
+                {
+                    _logger.LogError("Unable to auto track location: " + location.Name);
```
Use the overload that starts with the exception. It should print the message and stacktrace automatically.

MattEqualsCoder commented:
Made some bug fixes and added some additional autotracker detection of tricks and various situations. Also made Zelda tracking of deaths work a bit faster than before.

One larger feature I added is the ability to tell tracker "Hey tracker, look at this." Currently this only marks dungeon rewards after looking at the map, but I'm wanting to expand it to marking medallions as well as visible items. Only thing is to avoid either the player having to stop, tell tracker to look at it, then wait until they're sure that she picked it up correctly before moving, I basically save the action to perform upon seeing it. But, to make sure players are saying it at the right time, I have it clear out the action after 15 seconds. I figured this seemed like a reasonable compromise.

This also includes some of Fragger's json updates. At this point I'm not planning on making any additional changes outside of addressing pull request comments for this weekend.

MattEqualsCoder commented:
Oh, another thing worth noting is that there's a bit of a bug with the reward tracking with pendants. The SMZ3 generation code refers to pendants as either green or non-green. I actually tried to resolve this by adding separate red/blue pendant rewards, but sometimes they matched, sometimes they didn't. I'm hoping I can maybe figure that out, but I might add a non-green pendant rewarditem for this purpose if I can't.

Vivelin commented on src/Randomizer.SMZ3.Tracking/AutoTracking/EmulatorAction.cs:
```diff
@@ -0,0 +1,141 @@
+    /// <summary>
+    /// Which game(s) the message should be sent to the emulator in
+    /// </summary>
+    public enum Game
```
I know these enums are pretty tiny but please put these in separate files. The Game enum specifically could see use elsewhere (I know I added this in my MSU branch too)

Vivelin commented on src/Randomizer.SMZ3.Tracking/AutoTracking/USB2SNESConnector.cs:
```diff
@@ -0,0 +1,178 @@
+            }
+            catch (ObjectDisposedException e)
+            {
+                _logger.LogError(e, "Already disposed USB2SNES client");
```
You don't need to log this. This is kind of like getting a "File not found" error when you're trying to delete a file — the intended outcome was achieved either way.

TheRealFragger commented on src/Randomizer.SMZ3.Tracking/tracker.json:
```diff
@@ -2231,13 +2244,53 @@
+      "MockBall": [
+        "Good job on the tech. Was it a first try?",
+        "You just had to get these items early didn't you?",
+        "Do you know why I pulled you over? You were going 15 over the speed limit."
```
Perfect. I love it

Vivelin commented on src/Randomizer.SMZ3.Tracking/AutoTracking/AutoTracker.cs:
```diff
@@ -288,7 +288,7 @@ protected void Connector_MessageReceived(object? sender, EmulatorDataReceivedEve
         /// Sends requests out to the connected lua script
         /// </summary>
-        protected void SendMessages()
+        protected async void SendMessages()
```
...but you also shouldn't use `async void` unless you have no other choice (e.g. event handlers). So SendMessages should be `async Task` and Connector_Connected should be `async void` (because the event delegate specifies a void return type).

Async void still runs synchronously (because there is no `Task` to `await`)

MattEqualsCoder replied:
Huh. Didn't realize that. I'm surprised visual studio doesn't warn you about that reading up on it since it seems like events are the only general use case async void functions and can be really bad otherwise. 🤔 But sorry, I'll keep that in mind. I guess I need to brush up on all the complexities of C# tasks as they seem to be a consistent weak point of mine.

Vivelin replied:
I'm guessing that's because there's no way to see whether a method is an event handler from the method definition alone.

And yeah, async programming takes a while to get a good understanding of. I still don't fully understand all of the possible deadlocks and race conditions but I can recognize them at least. 

In general, you want to use `async Task`, `async Task<T>` or `async ValueTask<T>` where possible. `async IAsyncEnumerable<T>` is another fun new one.

MattEqualsCoder commented on src/Randomizer.SMZ3.Tracking/AutoTracking/AutoTracker.cs:
```diff
@@ -0,0 +1,864 @@
+        /// Tracks the current memory state of LttP for Tracker voice lines
+        /// </summary>
+        /// <param name="message">The message from the emulator with the memory state</param>
+        protected void CheckZeldaState(EmulatorAction action)
```
Yeah, I've had the same thoughts. I could possibly implement something like CS-Script and potentially a new config for all these where you basically put the if condition in there and what tracker can say as a result. This could allow people to add in their own potentially (especially if I add a debugger window that prints out the current Zelda/Metroid states).

Vivelin replied:
I'm not too sure about scripting because that could very easily lead to abuse.

I want to come up with something slightly complicated to add custom conditions to Tracker lines (e.g. only if go mode is enabled, or only if chat is connected). Whatever system that ends up being (probably something naive), I could probably apply to this, too.

MattEqualsCoder replied:
That's fair. I feel like that could get pretty unwieldy in and of itself with all these Zelda/Metroid state properties, but maybe I'm just not picturing it all in my head super well right now.
